With body function

BrandonRomano opened the pull request:
> This checkin adds a “WithBody” function and enables Wrecker to post
> JSON-encoded documents directly to a REST web service.
> 
> In the process, we’re simplifying the signature of the sendRequest()
> function by simply passing in the entire WreckerRequest object.

This is actually an in-project duplicate of #12, and will be favored as @benpate now has access to this repo.

Fixes #11

benpate committed:
Add WithBody() function (Issue #11)

This checkin adds a “WithBody” function and enables Wrecker to post
JSON-encoded documents directly to a REST web service.

In the process, we’re simplifying the signature of the sendRequest()
function by simply passing in the entire WreckerRequest object.

benpate committed:
[Issue #11] Test Cases

These test cases demonstrate wrecker working with JSON data in the
request body

BrandonRomano commented:
Hey @benpate... Just reviewed this PR -- looks great, fantastic work!

One thing I've noticed is that we're checking for nil value once before we call `prepareRequestBody` and again inside of `prepareRequestBody`...  Can we remove the check inside of the function? It's redundant and makes the code a little harder to read.

One thing I realized from this PR:

We're going to need to add a function so we can set the content-type at request level.  It looks like we're setting the content type to `application/json` when we call `prepareRequestBody`, but it looks like that function can actually handle more than just `application/json`.  If we call `request.WithBody("This is text").Execute()`, I imagine non-ideal things would happen!

So if you could:
- [ ] Review the PR, make sure I've not missed out anything in the rebase
- [ ] Remove (or respond to) the nil case in prepareRequestBody

After those two, feel free to merge... Again, fantastic work!

BrandonRomano committed:
Fixing tests

benpate commented:
Yes.  In the comments, I mentioned that this function was lifted directly from another MIT licensed project, so it could definitely be updated to fit Wrecker better.  I’ll make the changes you mentioned  to bring it in to line.

—
Ben

> On Jun 22, 2015, at 9:27 PM, Brandon Romano notifications@github.com wrote:
> 
> Hey @benpate https://github.com/benpate... Just reviewed this PR -- looks great, fantastic work!
> 
> One thing I've noticed is that we're checking for nil value once before we call prepareRequestBody and again inside of prepareRequestBody... Can we remove the check inside of the function? It's redundant and makes the code a little harder to read.
> 
> One thing I realized from this PR:
> 
> We're going to need to add a function so we can set the content-type at request level. It looks like we're setting the content type to application/json when we call prepareRequestBody, but it looks like that function can actually handle more than just application/json. If we call request.WithBody("This is text").Execute(), I imagine non-ideal things would happen!
> 
> So if you could:
> 
> Review the PR, make sure I've not missed out anything in the rebase
> Remove (or respond to) the nil case in prepareRequestBody
> After those two, feel free to merge... Again, fantastic work!
> 
> —
> Reply to this email directly or view it on GitHub https://github.com/BrandonRomano/wrecker/pull/14#issuecomment-114340696.

BrandonRomano commented:
Gotcha, yeah.  Thanks!

benpate committed:
Correctly JSON encode strings

removing alternate options in prepareRequestBody() function

benpate committed:
Test Cases

- adding test case for putting a JSON-encoded string
- moving all JSON test cases into their own file

benpate commented:
Updates and test cases posted.  Do you want to give this a once-over and merge it to master?

BrandonRomano commented:
Yeah let me check it

BrandonRomano commented on wrecker.go:
```diff
@@ -84,6 +104,20 @@ func (w *Wrecker) sendRequest(verb string, requestURL string, headers map[string
 	return resp, err
 }
+
+// prepareRequestBody() function was originally included in the
```
I would actually say we can remove this comment now as this function has deviated so much.

benpate committed:
updating comments

benpate commented:
done.

BrandonRomano commented on test/json_test.go:
```diff
@@ -0,0 +1,83 @@
+	status, ok := response.Content.(*string)
+
+	assert.True(t, ok)
+	assert.True(t, *status == "status code green")
```
So the test is passing, but I actually don't see how this is working now that we've removed support for text in the `prepareRequestBody` function... Hmm, perhaps you know why this is passing?

benpate replied:
Probably a lame unit test.  Will double check.

Sorry for the back-and-forth.

> On Jun 23, 2015, at 1:41 PM, Brandon Romano notifications@github.com wrote:
> 
> In test/json_test.go https://github.com/BrandonRomano/wrecker/pull/14#discussion_r33082635:
> 
> > -   response := models.Response{
> > -       Content: new(string),
> > -   }
> >   +
> > -   _, err := wreckerClient.Put("/status").
> > -       WithBody("status code green").
> > -       Into(&response).
> > -       Execute()
> >   +
> > -   assert.True(t, err == nil)
> > -   assert.True(t, response.Success)
> >   +
> > -   status, ok := response.Content.(*string)
> >   +
> > -   assert.True(t, ok)
> > -   assert.True(t, *status == "status code green")
> >   So the test is passing, but I actually don't see how this is working now that we've removed support for text in the prepareRequestBody function... Hmm, perhaps you know why this is passing?
> 
> —
> Reply to this email directly or view it on GitHub https://github.com/BrandonRomano/wrecker/pull/14/files#r33082635.

BrandonRomano replied:
It's fine -- this is how these usually go :stuck_out_tongue_winking_eye:

benpate replied:
The idea is to JSON-encode everything.  So, if you pass in an object, it gets encoded into a struct.  If you pass in a string, its simply surrounded with double quotes and escaped -- so JSON encoded as a string data.  

The server would still need to handle the response as JSON, but I think that's OK because we're sending the Content-Type as application/json.  In the future, it would be cool to enable other encoding types, like application/xml or even plain text.  But, those would be different encoders, and we'd need to build in a flag to switch between them.  

Does this answer make sense?  And, does this plan work for you?

BrandonRomano replied:
Okay yeah I understand now.  And yeah, that works for me... Definitely this is something I would like to do in the future!

BrandonRomano commented:
Okay @benpate I think this PR is ready to go :+1: !  Merging!
